substrate/client/service/src/client/client.rs

```diff
@@ -326,19 +325,35 @@ where
 			// dropped, the block will be unpinned automatically.
 			if let Some(ref notification) = finality_notification {
 				if let Err(err) = self.backend.pin_block(notification.hash) {
-					error!(
+					debug!(
 						"Unable to pin block for finality notification. hash: {}, Error: {}",
 						notification.hash, err
 					);
-				};
+				} else {
+					let _ = self
+						.unpin_worker_sender
+						.unbounded_send(UnpinWorkerMessage::AnnouncePin(notification.hash))
+						.map_err(|e| {
+							log::debug!(
+								"Unable to send AnnouncePin worker message for finality: {e}"
+							)
+						});
+				}
 			}
 
 			if let Some(ref notification) = import_notification {
 				if let Err(err) = self.backend.pin_block(notification.hash) {
-					error!(
+					debug!(
 						"Unable to pin block for import notification. hash: {}, Error: {}",
 						notification.hash, err
 					);
+				} else {
+					let _ = self
+						.unpin_worker_sender
+						.unbounded_send(UnpinWorkerMessage::AnnouncePin(notification.hash))
+						.map_err(|e| {
+							log::debug!("Unable to send AnnouncePin worker message for import: {e}")
+						});
```

Comment:
What about offloading the `backend.pin_block()` call to the `UnpinWorker` when it receives the `AccouncePin` message (which at this point would become `UnpinWorkerMessage::Pin`? 

In this way we completely delegate pins management to the Worker (which is also in charge of calling `unpin`. Also removes a couple of lines of duplicated logic from here.

Of course, in this case pinning becomes a "post-import" async action. Is that bearable?

Reply:
I think this would indeed be more elegant. 

> Of course, in this case pinning becomes a "post-import" async action. Is that bearable?

Exactly, that is the point of not doing it in the worker. I remember having some discussions about this and we want to make absolutely sure that the block is pinned before the notifications goes out to all listeners. So I would like to keep it this way.

Reply:
Yeah. I also see a couple of potential annoyances:

1) Any issue if we pin a block in the backend but the `unbounded_send` fails for any reason? 
2) A potential race condition: what if the `UnpinHandle::drop()` (-> and thus the `Unpin` message) is called before the worker receives the `AnnouncePin` message?  In this case maybe the pin counter in the worker will be :
   - first tried to be decremented (potentially decrementing the counter of another instance)
   - then in the best case will be re-incremented. But there is the possibility that the increment message will be lost as well

So 2 may introduce some nasty bugs maybe.
What about making pin request to the worker a sync message? So we wait before proceed sending out the import notification.

Reply:
Regarding 1:
I think it is not very realistic. It is an unbounded queue. Behind this channel is `async-channel`, and the docs state that sending can fail if the queue is full or the other side is closed. First one will not happen since its unbounded, closing could happen but means that the worker is dead.

Regarding 2:
Even if drop is called before the worker calls gets the `AnnouncePin` message, the drop message needs to reach the worker. I checked the `async-channel` docs again and sadly it does not seem to give any ordering guarantees. The channels from `tokio` however are mpsc and FIFO, so maybe I should switch to that. Would that address your concern?

> What about making pin request to the worker a sync message? So we wait before proceed sending out the import notification.

Hmm that would be an option. However I am more concerned that this would delay this rather critical codepath while holding the import lock. IMO that would be more concerning than the two edge cases above.

Reply:
> The channels from tokio however are mpsc and FIFO,

Well, I think would be a fair assumption for mpsc channels to be FIFO. But as you said I can't find any explicit statement on the doc. I asked https://github.com/smol-rs/async-channel/issues/6

> Hmm that would be an option. However I am more concerned that this would delay this rather critical codepath while holding the import lock. IMO that would be more concerning than the two edge cases above.

Fair enough.

Reply:
> Well, I think would be a fair assumption for mpsc channels to be FIFO.

I mean the async-channel ones are mpmc. My proposal was to switch to tokio channels because we only require a single consumer and they explicitly state it.

Reply:
I see. But would be interesting to understand how hey are working when there is a single consumer.
(the unpin worker is the only consumer right?)

Reply:
Yep, unpin worker ist the only one